### PR TITLE
US132923 [Adoption][Org Filter] Apply Tree Filter Selection API

### PR DESCRIPTION
**Pull request by Vitalii-Misechko**
[US132923](https://rally1.rallydev.com/#/?detail=/userstory/612183121391&fdp=true): [Adoption][Org Filter] Apply Tree Filter Selection API

* Functional Testing 
  * [x] Demo
  * [x] Test cases
    * [x] user drills into org unit then ou-filter shows only children of that org unit
      * [x] the root org unit or the parent org unit is not presented in the ou-filter list
    * [x] a user selects org units in the dropdown of ou-filter at any level, then the dashboard filters all charts except (System Access) according to the selection
    * [x] a user select org units of course level and then drill into course department level
      * [x] the org unit selection is cleared. So, a user sees that is nothing selected at a department level
      * [x] when a user drills up his selection is restored
    * [x] ou-filter is persisted in URL for the given drill
    * [x] ou-filter is restored from URL for the given drill
    * [x] There are no sums duplications
      * [x] System Access (not applicable, the data set is not filtered)
      * [x] Course Access Counts (not applicable, the root org unit is not present in API response)
      * [x] Tool Usage Trend by Tool  (not applicable, the root org unit is not present in API response)
      * [x] Tool Usage Trend by Role  (not applicable, the root org unit is not present in API response)
      * [x] Tool Usage Comparison  (not applicable, the root org unit is not present in API response)
      * [x] Enrollments and Withdrawals Trends  (not applicable, the root org unit is not present in API response)
      
### TODO in a separate PR

* Clear All doesn't cause a reload
* Show select all
* Drill that skips a few levels should clear all levels below the current drill

### CODE REVIEW

File: model/dataSets/toolUsageTrendByRoleDataSet.js [hunk 1]
```diff
@@ -18,15 +18,10 @@ export class ToolUsageTrendByRoleDataSet extends DataSet {
 		// rows now includes data for root org unit and all its children. For this card we only want to show root data
 		// *unless* we are drilling, which case we show all
 		const rows = (data?.toolUsageTrendData ?? [])
-			.filter(row => this._data.toolFilter.isSelected(row.t));
-
-		let rowsToAggregate = rows;
-		if (data?.lmsData?.isParentRoot) {
```
[the remark this answers is not in the document]
Comment (Vitalii-Misechko) [reply model/dataSets/toolUsageTrendByRoleDataSet.js:24]: > we want to show totals for all org units (not just visible ones) when there is no drill or selection.

> My intention here is to show all org unit data. I assumed that data for the root org unit is equivalent to the sum of all root's children.

Oh...I see. The original way data leaks. (A malicious user can reconstruct data by substructing from total values of visible children. I think we need to discuss this with PM/design

**Comment by johngwilkinson** [comment]
Testing

- combinations of date, drill, role, tool, ou filtering
- browser forward and back
- drill up and down after selecting

Slight bug - the department and course template here should show partially selected:
image

To reproduce, open up Accounting and it's course template and select a course. The template and department show "-" as expected. Now refresh the page. The selections are applied as expected, but the tree appears as in the above image. This only seems to happen when the page is loaded, not on hitting back or drilling up.

Also: I believe I was once able to get the org unit filter to show all children of root despite being drilled down. I haven't been able to reproduce, though.

Other than that, it's working so very well!

**Comment by Vitalii-Misechko** [comment]
> Slight bug - the department and course template here should show partially selected:
> To reproduce, open up Accounting and it's course template and select a course. The template and department show "-" as expected. Now refresh the page. The selections are applied as expected, but the tree appears as in the above image. This only seems to happen when the page is loaded, not on hitting back or drilling up.

Fixed the bug in this PR https://github.com/BrightspaceUILabs/ou-filter/pull/34

File: model/dataApiClient.js [hunk 2]
```diff
@@ -53,15 +53,15 @@ export class DataApiClient {
...
+	async _launchRequest({ selectedOrgUnitIds, userSettings, parentId, dateRange }) {
 		const url = new URL(concatMetronUrl(this._metronEndpoint, requestDataEndpoint));
 		const uri = url.toString();
 
 		const response = await d2lfetch.fetch(uri, {
 			method: 'POST',
 			headers: { 'cache-control': 'no-store', 'content-type': 'application/json' },
 			body: JSON.stringify({
-				selectedOrgUnitIdsCsv: orgUnitIds,
+				selectedOrgUnitIdsCsv: (selectedOrgUnitIds ?? []).join(','),
```
[the remark this answers is not in the document]
Comment (Vitalii-Misechko) [reply model/dataApiClient.js]: > Just noticing this. I think the bdp-query API should accept a JSON array of org unit ids, not a csv

Fixed

**Comment by johngwilkinson** [comment]
Rested with latest code here and on the server. Looks good!

**Review by johngwilkinson — APPROVED** [review]

**Pull request merged**